src/Ouroboros/Network/FetchLogic.hs

```diff
@@ -0,0 +1,1258 @@
+       -- fetch responses hit a soft timeout then it enters an aberrant state.
+       -- On a hard failure it is in the failure state, and it should be
+       -- removed as a candidate shortly thereafter.
+       --
+       peerFetchStatus :: PeerFetchStatus
+     }
+
+data PeerFetchStatus = PeerFetchStatusNormal
+                     | PeerFetchStatusAberrant
+                     | PeerFetchStatusFailure
+  deriving (Eq, Show)
+
+fetchRequestDecisions :: (HasHeader header, Ord peer)
+                      => FetchPolicyParams
+                      -> Map peer PeerFetchState
+                      -> [(ChainSuffix header, peer)]
+                      -> [FetchDecision header peer]
+fetchRequestDecisions fetchPolicyParams peerFetchStates =
+    go nConcurrentFetchPeers0 []
+  where
+    go _nConcurrentFetchPeers decisions [] = decisions
+    go  nConcurrentFetchPeers decisions ((chain, peer) : cps) =
+
+      let decision = fetchRequestDecision
+                       fetchPolicyParams
+                       nConcurrentFetchPeers
+                       peerFetchState
+                       chain
+                       peer
+
+          peerFetchState = peerFetchStates Map.! peer
+
+          nConcurrentFetchPeers'
+            -- increment if it was idle, and now will not be
+            | peerFetchBytesInFlight peerFetchState == 0
+            , Right{} <- decision = nConcurrentFetchPeers + 1
+            | otherwise           = nConcurrentFetchPeers
+
+      in go nConcurrentFetchPeers' (decision:decisions) cps
+
+    nConcurrentFetchPeers0 =
+      fromIntegral $ Prelude.length
+        [ () | s <- Map.elems peerFetchStates
+             , peerFetchReqsInFlight s > 0 ]
+
+
+fetchRequestDecision :: HasHeader header
+                     => FetchPolicyParams
+                     -> Word
+                     -> PeerFetchState
+                     -> ChainSuffix header
+                     -> peer
+                     -> FetchDecision header peer
+fetchRequestDecision FetchPolicyParams {
+                       maxInFlightBytesPerPeer,
+                       maxConcurrentFetchPeers
+                     }
+                     nConcurrentFetchPeers
+                     PeerFetchState{peerFetchBytesInFlight, peerFetchStatus}
+                     (ChainSuffix chain point) peer
+  | headPoint chain == point
+  = Left (FetchDeclineAllDownloaded peer point)
+
+{-
+    -- TODO: need to decide how this recovers and how the timeout is updated
+  | FetchActive _ _ TimeoutFired <- peerFetchStatus
+  = Left (FetchDeclineSlowPeer peer)
+
+--  | Just TimeoutCancelled <- mTimeoutState
+--  = 
+-}
+
+  | peerFetchBytesInFlight >= maxInFlightBytesPerPeer
+  = Left (FetchDeclineInFlightLimit peer maxInFlightBytesPerPeer)
+
+  | peerFetchBytesInFlight == 0
+  , nConcurrentFetchPeers >= maxConcurrentFetchPeers
+  = Left (FetchDeclineConcLimitLimit peer maxConcurrentFetchPeers)
+
```

Comment:
We're not covering the case, in which we already downloaded the blocks from some other peer.  Only the empty `ChainSuffix` case is covered.